server/src/main/java/org/opensearch/index/query/QueryRewriteContext.java

```diff
@@ -33,77 +33,40 @@
 
 import org.opensearch.client.Client;
 import org.opensearch.common.annotation.PublicApi;
-import org.opensearch.common.util.concurrent.CountDown;
 import org.opensearch.core.action.ActionListener;
 import org.opensearch.core.common.io.stream.NamedWriteableRegistry;
 import org.opensearch.core.xcontent.NamedXContentRegistry;
 import org.opensearch.core.xcontent.XContentParser;
 
-import java.util.ArrayList;
-import java.util.List;
 import java.util.function.BiConsumer;
-import java.util.function.LongSupplier;
 
 /**
  * Context object used to rewrite {@link QueryBuilder} instances into simplified version.
  *
  * @opensearch.api
  */
 @PublicApi(since = "1.0.0")
-public class QueryRewriteContext {
-    private final NamedXContentRegistry xContentRegistry;
-    private final NamedWriteableRegistry writeableRegistry;
-    protected final Client client;
-    protected final LongSupplier nowInMillis;
-    private final List<BiConsumer<Client, ActionListener<?>>> asyncActions = new ArrayList<>();
-    private final boolean validate;
-
-    public QueryRewriteContext(
-        NamedXContentRegistry xContentRegistry,
-        NamedWriteableRegistry writeableRegistry,
-        Client client,
-        LongSupplier nowInMillis
-    ) {
-        this(xContentRegistry, writeableRegistry, client, nowInMillis, false);
-    }
-
-    public QueryRewriteContext(
-        NamedXContentRegistry xContentRegistry,
-        NamedWriteableRegistry writeableRegistry,
-        Client client,
-        LongSupplier nowInMillis,
-        boolean validate
-    ) {
-
-        this.xContentRegistry = xContentRegistry;
-        this.writeableRegistry = writeableRegistry;
-        this.client = client;
-        this.nowInMillis = nowInMillis;
-        this.validate = validate;
-    }
-
+public interface QueryRewriteContext {
```

Comment:
This is changing a concrete class to an interface in a class marked `@opensearch.api` so it is not backwards compatible.  Existing calls in other code to `new QueryRewriteContext()` will fail.  

I don't see any in `opensearch-project` outside of the ones you've changed here in `OpenSearch`, but this breaks our API promise to any possible external plugins, etc.

Reply:
@msfroh and I had this discussion that for 2.19 or later if users are directly using this API, then they need to change to use BaseQueryRewriteContext. But it won't impact the elder version.

Reply:
I'd like to have a wider discussion on this issue.   SemVer has meanings and expectations.   I recognize there are a lot of fuzzy areas in OpenSearch but this class in particular is marked as API and I don't like to set the precedent of users being able to bump a minor version without their code depending on an API no longer compiling.

On other projects I've done a major version bump for things as simple as a method return value changing from an array to a List.  There seems to be a lot of reluctance to major version bumps here, but if we're going to say we're SemVer compliant we need to follow the rules.

CC: @andrross and @dblock for some input here.

Reply:
We will need to see what happens with the backward-compatibility check that runs on the 2.x branch.

I am of the opinion that semver applies to REST API compatibility (or more generally **user-facing** breaking changes) and developers of plugins should expect to make minor code changes within a major version, but that seems to be an unpopular opinion around here.